### ISSUE TITLE
Duplicating Datapoints in DataFlow

### ISSUE THREAD

**Issue by msbauer** [ppwwyyxx/tensorpack#207]
Hello,

I would like to duplicate datapoints in a DataFlow, so I wrote a ProxyDataFlow:
```
class RepeatedDataPoint(ProxyDataFlow):

    """ Take data points from another DataFlow and produce them a certain number of times
    dp1, ..., dp1, dp2, ..., dp2, ...
    """

    def __init__(self, ds, nr):
        """
        Args:
            ds (DataFlow): input DataFlow
            nr (int): number of times to repeat each datapoint.
        """
        self.nr = nr
        super(RepeatedDataPoint, self).__init__(ds)

    def size(self):
        """
        Raises:
            :class:`ValueError` when nr == -1.
        """
        return self.ds.size() * self.nr

    def get_data(self):
        for dp in self.ds.get_data():
            for _ in range(self.nr):
                yield dp
```

However, it does not seem to duplicate the datapoint at this point but also including augmentors applied afterwards:
```
ds = ... # shape (2000, 3000, 1)
ds = RepeatedDataPoint(ds, nr=5)
ds = AugmentImageComponent(ds, [imgaug.RandomCrop((256, 256))])
ds = PrintData(ds, num=5)
```
yields
```
datapoint 0<5 with 1 components consists of
   dp 0: is ndarray of shape (256, 256, 1) with range [-0.8151, 1.3173]
datapoint 1<5 with 1 components consists of
   dp 0: is ndarray of shape (256, 256, 1) with range [-0.8151, 1.3173]
datapoint 2<5 with 1 components consists of
   dp 0: is ndarray of shape (256, 256, 1) with range [-0.8151, 1.3173]
datapoint 3<5 with 1 components consists of
   dp 0: is ndarray of shape (256, 256, 1) with range [-0.8151, 1.3173]
datapoint 4<5 with 1 components consists of
   dp 0: is ndarray of shape (256, 256, 1) with range [-0.8151, 1.3173]
```
i.e. it always chooses the same patch because diffh==0 and diffw==0 after the first augmentation.

I presume I am doing something wrong. If so, please let me know what it is. Alternatively: Is there a better way to duplicate datapoints (I read large high-resolution files from disk. From each of them I want to extract many random crops. However, I don't want to re-read them every time.

Thanks
Matthias

**Comment by PatWie** [comment]
You need to backup the current point like
```
    for dp in d:
        dp_current = dp
        for _ in range(5):
            yield dp_current
```

**Comment by msbauer** [comment]
@PatWie: Thanks for the quick reply. Your proposed solution does not seem to work.

**Comment by PatWie** [comment]
One solution is patching the AugmentImageComponents class or using
```python
from tensorpack import *


ds = FakeData([(10, 10, 3)], random=False)
ds = MapData(ds, lambda dp: [dp[0], dp[0], dp[0]])
ds = AugmentImageComponent(ds, [imgaug.RandomCrop((4, 4))], index=0)
ds = AugmentImageComponent(ds, [imgaug.RandomCrop((4, 4))], index=1)
ds = AugmentImageComponent(ds, [imgaug.RandomCrop((4, 4))], index=2)
ds = PrintData(ds, num=5)
```


gives
```
datapoint 0<5 with 3 components consists of
   dp 0: is ndarray of shape (4, 4, 3) with range [0.0204, 0.9997]
   dp 1: is ndarray of shape (4, 4, 3) with range [0.0042, 0.9929]
   dp 2: is ndarray of shape (4, 4, 3) with range [0.0204, 0.9997]
datapoint 1<5 with 3 components consists of
   dp 0: is ndarray of shape (4, 4, 3) with range [0.0204, 0.9997]
   dp 1: is ndarray of shape (4, 4, 3) with range [0.0042, 0.9929]
   dp 2: is ndarray of shape (4, 4, 3) with range [0.0042, 0.9929]
datapoint 2<5 with 3 components consists of
   dp 0: is ndarray of shape (4, 4, 3) with range [0.0286, 0.9997]
   dp 1: is ndarray of shape (4, 4, 3) with range [0.0340, 0.9862]
   dp 2: is ndarray of shape (4, 4, 3) with range [0.0286, 0.9408]
datapoint 3<5 with 3 components consists of
   dp 0: is ndarray of shape (4, 4, 3) with range [0.0003, 0.9862]
   dp 1: is ndarray of shape (4, 4, 3) with range [0.0286, 0.9300]
   dp 2: is ndarray of shape (4, 4, 3) with range [0.0286, 0.9794]
datapoint 4<5 with 3 components consists of
   dp 0: is ndarray of shape (4, 4, 3) with range [0.0207, 0.9929]
   dp 1: is ndarray of shape (4, 4, 3) with range [0.0003, 0.9929]
   dp 2: is ndarray of shape (4, 4, 3) with range [0.0286, 0.9300]
```

**Comment by msbauer** [comment]
Thanks. I was also thinking about this workaround, however, then one has to chop up the datapoints again... can you explain to me what the problem with my original approach was?

**Comment by PatWie** [comment]
The problem is that AugmentImageComponents is implemented to
> Apply image augmentors on several components, with shared augmentation parameters.

**Comment by msbauer** [comment]
I am not using AugmentImageComponents but AugmentImageComponent (without s). That only applies it to the first component.

I meant more: why does my original RepeatedDataPoint not do what I expect it would do (which is: make 5 datapoints out of one datapoint which can be augmented independently)?

**Comment by PatWie** [comment]
What about:
````
from tensorpack import *
import copy

class RepeatedDataPoint(ProxyDataFlow):

    """ Take data points from another DataFlow and produce them a certain number of times
    dp1, ..., dp1, dp2, ..., dp2, ...
    """

    def __init__(self, ds, nr):
        """
        Args:
            ds (DataFlow): input DataFlow
            nr (int): number of times to repeat each datapoint.
        """
        self.nr = nr
        super(RepeatedDataPoint, self).__init__(ds)

    def size(self):
        """
        Raises:
            :class:`ValueError` when nr == -1.
        """
        return self.ds.size() * self.nr

    def get_data(self):
        for dp in self.ds.get_data():
            for _ in range(self.nr):
                yield copy.deepcopy(dp)


ds = FakeData([(10, 10, 3)], random=False)
ds = RepeatedDataPoint(ds, nr=5)
ds = AugmentImageComponents(ds, [imgaug.RandomCrop((4, 4))], index=[0])
ds = PrintData(ds, num=5)
````

**Comment by PatWie** [comment]
@ppwwyyxx I wonder why
`ds = AugmentImageComponents(ds, [imgaug.RandomCrop((4, 4))], index=[0], copy=True)`
from the latest version does not work here.

**Comment by ppwwyyxx** [comment]
You're right! My mistake. The copy option is not working.

**Comment by ppwwyyxx** [comment]
In #203 I was thinking about making copy=True the default. Maybe I should really do that since users can easily meet this problem..

**Comment by zxpeter** [comment]
Hi，
I'm also using FRCNN example and I'm changing get_train_dataflow() in data.py to get augmented training data for my own dataset (400). And I want to use filp and rotation to increase data num to 3*400=1200(origin, filp, rotation).So instead of putting shuffled dataflow with random filp or rotation in, I want to use 1200 as initial input for each epoch.


def get_train_dataflow():
    
    roidbs = DetectionDataset().load_training_roidbs()
    
    print_class_histogram(roidbs)
    num = len(roidbs)

    roidbs = list(filter(lambda img: len(img['boxes'][img['is_crowd'] == 0]) > 0, roidbs))
    logger.info("Filtered {} images which contain no non-crowd groudtruth boxes. Total #images for training: {}".format(
        num - len(roidbs), len(roidbs)))

    ds = DataFromList(roidbs, shuffle=True)
    ds = RepeatedData(ds, nr=3)

    aug = imgaug.AugmentorList(
            [CustomResize(cfg.PREPROC.TRAIN_SHORT_EDGE_SIZE, cfg.PREPROC.MAX_SIZE),
             imgaug.Flip(horiz=True)])
    aug_2 = imgaug.AugmentorList(
            [CustomResize(cfg.PREPROC.TRAIN_SHORT_EDGE_SIZE, cfg.PREPROC.MAX_SIZE),
             imgaug.Rotation(90)])

    def preprocess(roidb):
       ... ...

    if cfg.TRAINER == 'horovod':
        ds = MultiThreadMapData(ds, 5, preprocess)
    else:
        # ds = MultiProcessMapDataZMQ(ds, 10, preprocess)
        ds = MapData(ds, preprocess)

    return ds

My plan is use RepeatedData(ds, nr=3) to get duplicated dataflow and  then flip the second part and rotate the third part, but can't get those parts. I think it may have better solution for my problem, any help will be appreciated.

**Comment by ppwwyyxx** [comment]
@zxpeter You can either randomly choose between different augmentations or add a field in the extra `roidbs` to indicate whether it needs flip or rotation and read the extra field out in `preprocess`.

**Comment by nidetaoge** [comment]
@ppwwyyxx Sorry to bother, I want to try to use DataFromList to read (image_name, label) from txt list , for replacing dataset.ILSVRC12, like this:
    image_names, labels = load_names(datadir, 'train.txt')  ## read lines and split
    ds = DataFromList(list(zip(input_data, target_data)), shuffle=True)
    #ds = dataset.ILSVRC12(datadir, 'train', shuffle=True)
    ds = AugmentImageComponent(ds, augmentors, copy=False)
    def mapf(dp):
            fname, cls = dp
            im = cv2.imread(fname, cv2.IMREAD_COLOR)
            im = np.flip(im, axis=2)
            # print("fname={}".format(fname))
            im = aug.augment(im)
            return im, cls
    ds = MultiThreadMapData(ds, parallel, mapf, buffer_size=2000, strict=True)
    ds = BatchData(ds, batch_size, remainder=True)
    ds = PrefetchDataZMQ(ds, 1)

But I find that "list(zip(input_data, target_data))" is not the right operation, so how can I do it to use tensorpack to read from txt and use augment and prefrechdata again? Thanks so much for your help!
